deepspeed/__init__.py

```diff
@@ -66,6 +66,50 @@ def _parse_version(version_str):
 dist = None
 
 
+def _mark_initialized(trainobj: Union[torch.nn.Module, Optimizer, _LRScheduler]):
+    """Mark a trainobj as initialized by setting the ds_is_inited attribute to True."""
+    if hasattr(trainobj, 'ds_is_inited'):
+        assert trainobj.ds_is_inited, "Not expecting the training object has `ds_is_inited` to be False if it exists, make sure you didn't set it to False or called deepspeed.initialize on the model more than once."
+        return
+
+    trainobj.ds_is_inited = True
+
+
+def _is_initialized(trainobj: Union[torch.nn.Module, Optimizer, _LRScheduler]):
+    """Check if a trainobj has been initialized by checking the ds_is_inited attribute."""
+    if hasattr(trainobj, 'ds_is_inited'):
+        # we shouldn't hit the assert below, but just in case
+        assert trainobj.ds_is_inited, "Not expecting the training object has `ds_is_inited` to be False if it exists, make sure you didn't set it to False or called deepspeed.initialize on the model more than once."
+        return True
+    return False
+
+
+def _assert_trainobjs_not_inited(model: torch.nn.Module, optimizer: Optional[Optimizer],
+                                 lr_scheduler: Optional[_LRScheduler]):
+    """Enforce the model, optimizer, and lr_scheduler have not been used in a previous deepspeed.initialize call."""
+    if _is_initialized(model):
+        raise ValueError(
+            "Model has already been initialized, please make sure to only call deepspeed.initialize on a model once.")
+    if optimizer is not None and _is_initialized(optimizer):
+        raise ValueError(
+            "Optimizer has already been initialized, please make sure to only call deepspeed.initialize on an optimizer once."
+        )
+    if lr_scheduler is not None and _is_initialized(lr_scheduler):
```

Comment:
Ditto for lr_scheduler
https://github.com/microsoft/DeepSpeed/blob/4cd1d97460b677563d57f07a293724bdc02e0ef5/deepspeed/__init__.py#L74